rimage/image_file.go

```diff
@@ -209,19 +206,34 @@ func EncodeImage(ctx context.Context, img image.Image, mimeType string) ([]byte,
 	_, span := trace.StartSpan(ctx, "rimage::EncodeImage::"+mimeType)
 	defer span.End()
 
-	actualOutMIME, _ := ut.CheckLazyMIMEType(mimeType)
-	if lazy, ok := img.(*LazyEncodedImage); ok && lazy.MIMEType() == actualOutMIME {
-		// fast path - zero copy
-		return lazy.RawData(), nil
+	if lazyImg, ok := img.(*LazyEncodedImage); ok {
+		return lazyImg.imgBytes, nil
 	}
 
 	var buf bytes.Buffer
 	bounds := img.Bounds()
-	switch actualOutMIME {
+	switch mimeType {
 	case ut.MimeTypeRawRGBA:
-		imgCopy := image.NewNRGBA64(bounds)
-		draw.Draw(imgCopy, bounds, img, bounds.Min, draw.Src)
-		buf.Write(imgCopy.Pix)
+		// Here we create a custom header to prepend to Raw RGBA data. Credit to
+		// Ben Zotto for inventing this formulation
+		// https://bzotto.medium.com/introducing-the-rgba-bitmap-file-format-4a8a94329e2c
+		buf.Write(RGBABitmapMagicNumber)
+		widthBytes := make([]byte, 4)
+		heightBytes := make([]byte, 4)
+		binary.BigEndian.PutUint32(widthBytes, uint32(bounds.Dx()))
+		binary.BigEndian.PutUint32(heightBytes, uint32(bounds.Dy()))
+		buf.Write(widthBytes)
+		buf.Write(heightBytes)
+		imgStruct, ok := img.(*image.NRGBA)
+		if !ok {
+			_, ok := img.(*Image)
+			if !ok {
+				return nil, errors.Errorf("underlying struct for image not NRGBA or rimage.Image, detected %T", img)
+			}
```

Comment:
You don't need to error here. draw.Draw is smart. It can take any image that fulfills the image.Image interface, and write it properly into `imgStruct`. So rather than erroring, just let draw.Draw do its thing.

Reply:
If you wanted to, you don't even need to do a check for underlying type. draw.Draw can take care of it.